### PR TITLE
In memory model transfer between neuron and coreneuron

### PR DESCRIPTION
**Pull request by pramodk**
- Merging first draft from Michael to transfer neuron model to coreneuron with DMA

### CODE REVIEW

File: coreneuron/nrniv/global_vars.cpp [hunk 1]
```diff
@@ -9,6 +9,10 @@
 #include "coreneuron/nrniv/nrniv_decl.h"
 #include "coreneuron/nrnoc/membfunc.h"
 #include "coreneuron/nrniv/nrn_assert.h"
+#include "coreneuron/nrniv/nrn2core_direct.h"
+
+void* (*nrn2core_get_global_dbl_item_)(void*, const char*& name, int& size, double*& val);
+int (*nrn2core_get_global_int_item_)(const char* name);
```
Comment (pramodk) [inline coreneuron/nrniv/global_vars.cpp]: can we transfer these to single header? (nrn2core_direct.h)

File: coreneuron/nrniv/main1.cpp [hunk 2]
```diff
@@ -54,9 +55,78 @@ THE POSSIBILITY OF SUCH DAMAGE.
...
+        int sp = isspace(arg[i]);
+        if (inarg && sp) {  // start whitespace following previous arg
+            inarg = 0;
+        } else if (!inarg && !sp) {  // first char of current arg
+            argc++;
+            inarg = 1;
+        }
+    }
+    argc += use_mpi ? 2 : 1;  // corenrn -mpi or just corenrn
+    char** argv = new char*[argc];
```
Comment (pramodk) [inline coreneuron/nrniv/main1.cpp]: corresponding delete missing?

File: coreneuron/nrniv/main1.cpp [hunk 3]
```diff
@@ -54,9 +55,78 @@ THE POSSIBILITY OF SUCH DAMAGE.
...
+
+    // recount and fill argv
+    argc = 0;
+    argv[argc++] = strdup("corenrn");
+    if (use_mpi) {
+        argv[argc++] = strdup("-mpi");
```
Comment (pramodk) [inline coreneuron/nrniv/main1.cpp]: corresponding free missing?

File: coreneuron/nrniv/main1.cpp [hunk 4]
```diff
@@ -54,9 +55,78 @@ THE POSSIBILITY OF SUCH DAMAGE.
...
+    }
+    inarg = 0;
+    int first = 0;
+    int i = 0;
+    do {
+        char c = arg[i];
+        int sp = isspace(c) ? 1 : ((c == '\0') ? 1 : 0);
+        if (inarg && sp) {  // start whitespace following previous arg
+            inarg = 0;
+            argv[argc++] = strndup(arg + first, i - first);
```
Comment (pramodk) [inline coreneuron/nrniv/main1.cpp]: corresponding free missing?

File: coreneuron/nrniv/main1.cpp [hunk 5]
```diff
@@ -54,9 +55,78 @@ THE POSSIBILITY OF SUCH DAMAGE.
...
+        } else if (!inarg && !sp) {  // first char of current arg
+            inarg = 1;
+            first = i;
+        }
+    } while (arg[i++]);
+
+    printf("arg: %s\n", arg);
+    for (i = 0; i < argc; ++i) {
+        printf("%d %s\n", i, argv[i]);
+    }
```
Comment (pramodk) [inline coreneuron/nrniv/main1.cpp]: only rank 0 should print?

File: coreneuron/nrniv/mk_mech.cpp [hunk 6]
```diff
@@ -62,30 +66,73 @@ int nrn_need_byteswap;
 
 std::map<std::string, int> mech2type;
 
+extern "C" {
+void (*nrn2core_mkmech_info_)(std::ostream&);
+}
```
Comment (pramodk) [inline coreneuron/nrniv/mk_mech.cpp]: can this be moved to nrn2core_direct.h?

File: coreneuron/nrniv/nrn_setup.cpp [hunk 7]
```diff
@@ -49,6 +49,85 @@ THE POSSIBILITY OF SUCH DAMAGE.
...
+
+int (*nrn2core_get_dat2_vecplay_inst_)(int tid,
+                                       int i,
+                                       int& vptype,
+                                       int& mtype,
+                                       int& ix,
+                                       int& sz,
+                                       double*& yvec,
+                                       double*& tvec);
+
```
Comment (pramodk) [inline coreneuron/nrniv/nrn_setup.cpp]: if these are declared in nrn2core_direct.h, why we need to declare here again?

**Comment by pramodk** [comment]
I think we should run valgrind to make sure all memory is getting freed in embedded mode.

File: coreneuron/nrniv/global_vars.cpp (hunk 1 above)
Comment (nrnhines) [reply coreneuron/nrniv/global_vars.cpp]: I thought they needed a separare declaration in the header (extern) and then a definition in
some file (memory to hold the function pointer)

File: coreneuron/nrniv/main1.cpp (hunk 3 above)
Comment (nrnhines) [reply coreneuron/nrniv/main1.cpp]: Yes.  In general, though, on return from nrncore_run it is not clear at the moment how much access NEURON needs to the coreneuron results. .eg. perhaps there should be an inverse
to nrnbbcore_write. Or, as we mentioned in the paper a series of accessor functions of various grain.

File: coreneuron/nrniv/main1.cpp (hunk 4 above)
Comment (nrnhines) [reply coreneuron/nrniv/main1.cpp]: It is safe to free all argv and argv itself on return from nrncore_run.

File: coreneuron/nrniv/main1.cpp (hunk 5 above)
Comment (nrnhines) [reply coreneuron/nrniv/main1.cpp]: yes. unfortunately, coreneuron has not started mpi yet and nrniv did not pass its version of
nrnmpi_myid to nrncore_run

File: coreneuron/nrniv/mk_mech.cpp (hunk 6 above)
Comment (nrnhines) [reply coreneuron/nrniv/mk_mech.cpp]: I thought it needed a unique definition

File: coreneuron/nrniv/nrn_setup.cpp (hunk 7 above)
Comment (nrnhines) [reply coreneuron/nrniv/nrn_setup.cpp]: they are all extern. the file is included in multiple places.

**Comment by nrnhines** [comment]
With respect to valgrind, that is an excellent idea. Note that the ability to make multiple calls to
pc.nrncore_run(string) has not been thought through yet.

File: coreneuron/nrniv/main1.cpp (hunk 5 above)
Comment (nrnhines) [reply coreneuron/nrniv/main1.cpp]: These were for debugging and can be removed

**Comment by pramodk** [comment]
@nrnhines : with last two commits I have fixed memory leaks with valgrind. I am not sure about removing event from queue:

```
==53751== 96 bytes in 2 blocks are definitely lost in loss record 79 of 118
==53751==    at 0x100099616: malloc (in /usr/local/Cellar/valgrind/3.13.0/lib/valgrind/vgpreload_memcheck-amd64-darwin.so)
==53751==    by 0x10135BE0D: operator new(unsigned long) (in /usr/lib/libc++abi.dylib)
==53751==    by 0x1000F668B: coreneuron::TQueue<(coreneuron::container)1>::insert(double, void*) (tqueue.ipp:243)
==53751==    by 0x1000F5294: coreneuron::NetCvode::event(double, coreneuron::DiscreteEvent*, coreneuron::NrnThread*) (netcvode.cpp:251)
==53751==    by 0x100101331: coreneuron::NetParEvent::send(double, coreneuron::NetCvode*, coreneuron::NrnThread*) (netpar.cpp:132)
==53751==    by 0x100101BEF: coreneuron::nrn_spike_exchange_init() (netpar.cpp:306)
==53751==    by 0x10011C69E: coreneuron::checkpoint_initialize() (nrn_checkpoint.cpp:860)
==53751==    by 0x1000D3F7A: solve_core (main1.cpp:396)
==53751==    by 0x100001F81: main (coreneuron.cpp:33)
```

Obviously this is not related to this branch though.

**Pull request merged**